### PR TITLE
[FLINK-7574][build] POM Cleanup flink-clients

### PR DESCRIPTION
**Pull request by yew1eb**
Re-created #4712.

## What is the purpose of the change

This PR changes the `flink-clients` pom to

-  contain unused dependencies
-  contain all used dependencies

## Brief change log

- *Define the `maven-dependency-plugin` plugin version in root POM pluginManagement*
- *Enable `dependency:analyze-only` at root POM plugins*
- *Disable the dependency analysis of flink submodules, except the `flink-clients`*
- *Cleanup`flink-clients` dependencies(used but undeclared / declared but unused)*
- *[travis] Only execute `dependency:analyze-only` in "Misc" profile*

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (**yes** / no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

### CODE REVIEW

**Comment by yew1eb** [comment]
`mvn clean verify -pl flink-clients`, No dependency problems found.

> [INFO] --- maven-dependency-plugin:3.0.2:analyze-only (analyze) @ flink-clients_2.11 ---
> [INFO] No dependency problems found
> [INFO] ------------------------------------------------------------------------
> [INFO] BUILD SUCCESS
> [INFO] ------------------------------------------------------------------------
> [INFO] Total time: 17.457 s
> [INFO] Finished at: 2017-11-28T12:59:50+08:00
> [INFO] Final Memory: 76M/720M
> [INFO] ------------------------------------------------------------------------
>

File: pom.xml [hunk 1]
```diff
@@ -891,6 +905,41 @@ under the License.
 
 	<build>
 		<plugins>
+			<plugin>
+				<groupId>org.apache.maven.plugins</groupId>
+				<artifactId>maven-dependency-plugin</artifactId>
+				<version>3.0.2</version>
```
Comment (greghogan) [inline pom.xml]: Redundant from the `<pluginManagement>` section.

File: pom.xml [hunk 2]
```diff
@@ -891,6 +905,41 @@ under the License.
...
+						<id>analyze</id>
+						<goals>
+							<goal>analyze-only</goal>
+						</goals>
+						<configuration>
+							<!-- Ignore runtime-only dependencies in analysis -->
+							<ignoreNonCompile>true</ignoreNonCompile>
+							<outputXML>true</outputXML>
+							<failOnWarning>true</failOnWarning>
+							<ignoredUnusedDeclaredDependencies>
```
Comment (greghogan) [inline pom.xml]: Would it be better to move the dependencies out of the parent pom?

Comment (zentol) [reply pom.xml]: I would keep logging, junit, jsr305 and force-shading in the parent pom as they are used by virtually all modules and would just clutter up other poms.

But I wouldn't mind moving the mocking/hamcrest dependencies to child poms.

**Comment by zentol** [comment]
These should generally be pretty safe changes to make IMO, and well worth it. In particular the undeclared usage of transitive dependencies (that may be shaded) frequently caused headaches, and should now be much easier to detect.

We have to look out for dependencies that are used exclusively via reflection and those we have for the sake of including them in flink-dist; can't think of other special cases at the moment.

**Comment by zentol** [comment]
We also have to look out for dependencies that were added to include a transitive dependency in a fat jar, like we do for reporters/connectors,

**Comment by yew1eb** [comment]
Hi @greghogan,  You're quite right, Thanks for your advice.  
For a discussion of this improvement, you can see #4712.

File: pom.xml (hunk 1 above)
Comment (yew1eb) [reply pom.xml]: will remove.

**Comment by yew1eb** [comment]
@zentol , Thanks for your suggestions, i will be very careful to clean up the dependencies.
I will start with the top modules (not used by other modules), order as follow:
> ├── flink-clients
> ├── flink-scala-shell
> ├── flink-end-to-end-tests
> ├── flink-fs-tests
> ├── flink-mesos
> ├── flink-yarn-tests
> ├── flink-yarn
> ├── flink-formats
> │   ├── flink-avro
> ├── flink-java8
> ├── flink-runtime-web
> ├── flink-examples
> │   ├── flink-examples-batch
> │   ├── flink-examples-streaming
> │   ├── flink-examples-table
> ├── flink-libraries
> │   ├── flink-cep
> │   ├── flink-cep-scala
> │   ├── flink-gelly
> │   ├── flink-gelly-examples
> │   ├── flink-gelly-scala
> │   ├── flink-ml
> │   ├── flink-python
> │   ├── flink-table
> ├── flink-metrics
> │   ├── flink-metrics-core
> │   ├── flink-metrics-datadog
> │   ├── flink-metrics-dropwizard
> │   ├── flink-metrics-ganglia
> │   ├── flink-metrics-graphite
> │   ├── flink-metrics-jmx
> │   ├── flink-metrics-prometheus
> │   ├── flink-metrics-slf4j
> │   ├── flink-metrics-statsd
> ├── flink-contrib
> │   ├── flink-connector-wikiedits
> │   ├── flink-statebackend-rocksdb
> │   ├── flink-storm
> │   ├── flink-storm-examples
> │   ├── flink-streaming-contrib
> │   ├── flink-tweet-inputformat
> ├── flink-filesystems
> │   ├── flink-hadoop-fs
> │   ├── flink-mapr-fs
> │   ├── flink-s3-fs-hadoop
> │   ├── flink-s3-fs-presto
> ├── flink-connectors
> │   ├── flink-connector-cassandra
> │   ├── flink-connector-elasticsearch
> │   ├── flink-connector-elasticsearch-base
> │   ├── flink-connector-elasticsearch2
> │   ├── flink-connector-elasticsearch5
> │   ├── flink-connector-filesystem
> │   ├── flink-connector-kafka-0.10
> │   ├── flink-connector-kafka-0.11
> │   ├── flink-connector-kafka-0.8
> │   ├── flink-connector-kafka-0.9
> │   ├── flink-connector-kafka-base
> │   ├── flink-connector-kinesis
> │   ├── flink-connector-nifi
> │   ├── flink-connector-rabbitmq
> │   ├── flink-connector-twitter
> │   ├── flink-hadoop-compatibility
> │   ├── flink-hbase
> │   ├── flink-hcatalog
> │   ├── flink-jdbc
> │   ├── flink-orc
> ├── flink-optimizer
> ├── flink-queryable-state
> ├── flink-test-utils-parent
> │   ├── flink-test-utils
> │   ├── flink-test-utils-junit
> ├── flink-tests
> ├── flink-scala
> ├── flink-streaming-java
> ├── flink-streaming-scala
> ├── flink-runtime
> ├── flink-java
> ├── flink-core

File: pom.xml [hunk 3]
```diff
@@ -581,6 +581,20 @@ under the License.
 				<version>0.9.10</version>
 				<scope>test</scope>
 			</dependency>
+
+			<dependency>
+				<groupId>org.powermock</groupId>
+				<artifactId>powermock-core</artifactId>
```
Comment (zentol) [inline pom.xml]: these should also be removed, no?

Comment (yew1eb) [reply pom.xml]: removed.

**Comment by yew1eb** [comment]
@zentol thanks for your review. I have updated the PR.


`mvn clean verify -Dverbose -pl flink-clients`，No dependency problems found:

```
[INFO] --- maven-dependency-plugin:3.0.2:analyze-only (analyze) @ flink-clients_2.11 ---
[INFO] Used declared dependencies found:
[INFO]    org.apache.flink:flink-core:jar:1.5-SNAPSHOT:compile
[INFO]    org.apache.flink:flink-runtime_2.11:jar:1.5-SNAPSHOT:compile
[INFO]    org.apache.flink:flink-optimizer_2.11:jar:1.5-SNAPSHOT:compile
[INFO]    org.apache.flink:flink-java:jar:1.5-SNAPSHOT:compile
[INFO]    org.apache.flink:flink-shaded-netty:jar:4.0.27.Final-2.0:compile
[INFO]    commons-cli:commons-cli:jar:1.3.1:compile
[INFO]    org.scala-lang:scala-library:jar:2.11.11:compile
[INFO]    com.typesafe.akka:akka-actor_2.11:jar:2.4.20:compile
[INFO]    org.apache.flink:flink-test-utils-junit:jar:1.5-SNAPSHOT:test
[INFO]    org.apache.flink:flink-runtime_2.11:test-jar:tests:1.5-SNAPSHOT:test
[INFO]    org.slf4j:slf4j-api:jar:1.7.7:compile
[INFO]    com.google.code.findbugs:jsr305:jar:1.3.9:compile
[INFO]    junit:junit:jar:4.12:test
[INFO]    org.mockito:mockito-all:jar:1.10.19:test
[INFO]    org.powermock:powermock-api-mockito:jar:1.6.5:test
[INFO] Ignored unused declared dependencies:
[INFO]    org.apache.flink:force-shading:jar:1.5-SNAPSHOT:compile
```

**Comment by zentol** [comment]
merging.

File: pom.xml (hunk 2 above)
Comment (StephanEwen) [reply pom.xml]: Can we keep the root dependencies? I find the current way actually convenient and avoiding pom clutter.

As Chesnay mentioned, the compile scope dependencies should be okay (used everywhere).

Could the test dependencies be separately skipped?

**Comment by StephanEwen** [comment]
Is this merged already?

I think the idea is nice, but I would say let's merge this if we have commitment to follow up on this, meaning fixing the other modules.

We already have a partially done dependency convergence work in the pom files, so I would vote to add another partially done dependency analysis work only if there is a commitment for timely followup and conclusion of this work (for example before 1.5 release, feature freeze targeted end of January).

What do you think?

**Comment by zentol** [comment]
Not merged yet,

**Comment by zentol** [comment]
I think it would be fine to start merging things now since the changes are more isolated than the dependency convergence one.

**Comment by zentol** [comment]
We will have to redo this from scratch since too much time has passed.

**Pull request closed without merging**